File: LAB_7_C#_C++/FloatList.cpp
#include "FloatList.h"
#include <algorithm>

float& FloatList::operator [] (const int index)
{
	return floatlist[index];
}

int FloatList::Length()
{
	return count;
}

FloatList::FloatList()
{
	count = 0;
}

void FloatList::Add(float num)
{
    float* newList = new float[count + 1];

    count++;
    newList[0] = num;
            
    for(int i = 1; i < count; i++)
    {
        newList[i] = floatlist[i - 1];
    }

    floatlist = newList;
}

int FloatList::Find(float num)
{
    for (int i = 0; i < count; i++)
        if (num == floatlist[i])
            return i;
    return -1;
}

void FloatList::Remove(float num)
{
    int index = Find(num);

    if (index == -1)
        return;

    float* newList = new float[--count];

    for (int i = 0; i < index; i++)
        newList[i] = floatlist[i];

    for (int i = index; i < count; i++)
        newList[i] = floatlist[i + 1];

    floatlist = newList;
}

int FloatList::BiggerThanAverage()
{
    if (count < 2)
        return 0;

    int amount = 0;
    float average;
    float* list = new float[count];

	for (int i = 0; i < count; i++)
		list[i] = floatlist[i];

	int n = sizeof(list) / sizeof(list[0]);

	std::sort(list, list + n);

    average = list[count / 2];

	for(int i = 0; i < count; i++)
		if (list[i] > average)
			amount++;

    return amount;
}

void FloatList::DeleteAllNegativeNums()
{
    for(int i = 0; i < count; i++)
    {
        if (floatlist[i] < 0)
        {
            Remove(floatlist[i]);
            i--;
        }
    }
}

File: LAB_7_C#_C++/FloatList.cs
using System;
using System.Collections;


namespace OOP_LAB7
{
    public class FloatList : IEnumerable  
    {
        Node head; // first elem
        Node tail; // last elem
        public int Count { get; private set; }
        public bool IsEmpty { get { return Count == 0; } }

        // Adding new element at the end of list
        public void Add(float data)
        {
            Node node = new Node(data);
            node.Next = head;
            head = node;

            if (Count == 0)
                tail = head;

            Count++;
        }
        // Removing element
        public bool Remove(float data)
        {
            Node current = head;
            Node previous = null;

            while (current != null)
            {
                if (current.Data.Equals(data))
                {
                    
                    if (previous != null)
                    {
                        // убираем узел current, теперь previous ссылается не на current, а на current.Next
                        previous.Next = current.Next;

                        // Если current.Next не установлен, значит узел последний,
                        // изменяем переменную tail
                        if (current.Next == null)
                            tail = previous;
                    }
                    else
                    {
                        // если удаляется первый элемент
                        // переустанавливаем значение head
                        head = head.Next;

                        // если после удаления список пуст, сбрасываем tail
                        if (head == null)
                            tail = null;
                    }
                    Count--;
                    return true;
                }

                previous = current;
                current = current.Next;
            }
            return false;
        }

        public void Clear()
        {
            head = null;
            tail = null;
            Count = 0;
        }

        public bool Contains(float data)
        {
            Node current = head;
            while (current != null)
            {
                if (current.Data.Equals(data))
                    return true;
                current = current.Next;
            }
            return false;
        }

        public int BiggerThanAverage()
        {
            float average = 0;
            float sum = 0;
            int amount = 0;

            Node current = head;

            while(current != null)
            {
                sum += current.Data;
                current = current.Next;
            }
            average = sum / Count;

            current = head;
            while (current != null)
            {
                if (current.Data > average)
                    amount++;
                current = current.Next;
            }

            return amount;
        }

        public void DeleteAllNegativeNums()
        {
            Node current = head;

            for(int i = 0; i < Count; i++)
            {
                if (current.Data < 0)
                {
                    Remove(current.Data);
                    i--;
                }

                current = current.Next;
            }
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            Node current = head;
            while (current != null)
            {
                yield return current.Data;
                current = current.Next;
            }
        }
    }
}
